add `granular_approval` flag to `approve` plugin

ykakarap opened the pull request:
Context for the approve2 plugin:
- _TODO: add links here_

This is a WIP PR.  
More details on how the plugin works will be added soon. 

This PR needs a KEP for the new `approve2` plugin.

**TLDR' on `approve2` plugin:**
This plugin is an updated approve plugin that supports
granular approvals and regex in the OWNERS files

This commit adds a new PR status section to the approval notifier message.
This section shows the status of approvals of the folders involved in the PR.
It shows if all the files under a folder are completely approved, partially approved
or unapproved. The folders used are the ones where the owners files associated
with the original PR reside.

alvaroaleman commented:
@ykakarap thanks for your work! Unfortunately, a 7k LOC PR is very time-consuming to review.

Could you please try to de-duplicate the tests and run them against both versions, denoting expected failures in the current approval plugin as applicable?

alvaroaleman commented:
> Could you please try to de-duplicate the tests and run them against both versions, denoting expected failures in the current approval plugin as applicable?

Or alternatively, maybe one commit that just copies the old approval plugin, then another commit with your changes on top of that?

ykakarap commented:
>Or alternatively, maybe one commit that just copies the old approval plugin, then another commit with your changes on top of that?

I will do this 🙂 
Will make the review process easier.

ykakarap commented:
Hey @alvaroaleman, I split the PR into 2 commits. The first commit simply duplicates the old `approve` as an `approve2` plugin. 
The second commit has the all the changes that make up the new plugin, including registration of handlers and integration.

Hope this helps with the review. 🙂 

PS: I am also working on a KEP that goes with this. This PR will probably remain as WIP till the KEP is hashed out.

/cc @nikhita

nikhita commented:
/cc @mrbobbytables @spiffxp

BenTheElder commented:
While rolling this out eventually likely needs review from contribex, I don't think this has to do a KEP.

I think we can make some pretty small changes comparatively based on the previous sig testing meeting discussion:
1. Put this in the approve plugin, gate it behind a config option instead of configuring as a different plugin name entirely, but it can and should still have distinct codepaths for the logic
2. Drop the `2` from the command regexes, for a more natural transition as repos phase in/out of enabling it.

Otherwise just needs cleanup and PR review. I know some people want this quite badly!

ykakarap commented:
@BenTheElder ack!

I will be back working on this in the next few weeks.

The 2 points you made are

ykakarap commented:
While doing the clean up I will also address the test pipeline failures on this PR.

ykakarap commented:
/retest

ykakarap commented:
@BenTheElder @spiffxp  Hi 👋🏼 
I have addressed both the comments on the PR and the comments I received from the SIG Testing meeting. 
* I moved the granular approval behavior behind a `granular_approval` flag in the `approve` plugin and dropped the use of a separate `approve2` plugin.
  * I found that the easiest and cleanest way to do this was to keep the granular approval variant of the approval plugins logic in a separate package and different from the original plugin. This is because the granular approval has a lot of changes and would introduce duplicates of almost all the functions in the `approve/approvers` package. Hence it seemed best to keep the behavior in separate packages. Hence granular approval logic sits in `approve/approvers2` package.
* The second feedback point was to show the list of files that are yet to be approved and are already approved. The feedback was that having just the number of files that are yet  to be approved might not give the full picture. The "status of the PR" section now will also show the list of files that are yet to be approved and approved (approved files are striked out). Please look at the below screenshot for an example: 
<img width="856" alt="Screen Shot 2021-08-04 at 9 37 00 AM" src="https://user-images.githubusercontent.com/22120431/128220966-fd2ef64d-7f7c-4d16-97f1-9013d35d7773.png">

The files list is under a collapsable section that is closed by default (the `pkg/registry` section in the above example was manually expanded). This is to have a cleaner look and avoid huge comments for large PRs.

PS: This PR does not change the configs of the approve plugin to use granular approvals for any of the repos. I was planning on rolling out this feature to only to 1 or 2 repos (is the k/test-infra repo a good candidate?) in the beginning to collect user feedback. Any advice on a different roll-out strategy is welcome 🙂 .

cc @dims @alvaroaleman @nikhita

spiffxp commented:
Pinging for attention. I'd like to see us roll this out after enhancements freeze

ykakarap commented:
/retest

BenTheElder commented:
we discussed this in today's SIG meeting but honestly I've been a bit out of it and don't recall the outcome clearly ... cc @spiffxp @alvaroaleman

ykakarap commented:
/remove-lifecycle stale

BenTheElder commented:
so this will need a rebase, but before we ask OP to do so, is this something we're willing to move forward with?
cc @stevekuznetsov @alvaroaleman @chaodaiG @cjwagner

cjwagner commented:
> so this will need a rebase, but before we ask OP to do so, is this something we're willing to move forward with? cc @stevekuznetsov @alvaroaleman @chaodaiG @cjwagner

The idea SGTM assuming we have interest in this feature, but I don't think I can personally commit the time to reviewing this PR now.

BenTheElder commented on prow/plugins/approve/BUILD.bazel:
```diff
@@ -12,6 +12,7 @@ go_library(
         "//prow/pluginhelp:go_default_library",
         "//prow/plugins:go_default_library",
         "//prow/plugins/approve/approvers:go_default_library",
+        "//prow/plugins/approve/approvers2:go_default_library",
```
these bazel files can be deleted now

BenTheElder reviewed (commented):
/lint

@chaodaiG maybe?

MadhavJivrajani commented:
/cc @michelle192837

BenTheElder commented on prow/plugins/approve/approve.go:
```diff
@@ -730,3 +663,241 @@ func getLast(cs []*comment) *comment {
+	updateNoficiation(linkURL *url.URL, commandHelpLink, prProcessLink, org, repo, branch string, latestNotification *comment) *string
+}
+
+func NewApproveHelper(opts *plugins.Approve, log *logrus.Entry, filenames []string, repo approvers.Repo, seed int64) approveHelper {
```
These comments apply to the existing code and probably should just be ignored for now.

MadhavJivrajani commented:
/test pull-test-infra-unit-test

Not able to reproduce the unit test failure locally and the logs also seem to have been cleaned up.

MadhavJivrajani commented:
yep test failures are due to the merge conflict. 
@ykakarap once this is rebased, the failures should go away!

BenTheElder commented on gcsweb/cmd/gcsweb/templates.go:
```diff
@@ -119,9 +119,9 @@ var tmplContentFooter = template.Must(template.New("content-footer").Parse(tmplC
+		Path string
 	}{
-		Path:    path,
+		Path: path,
```
unrelated diff?

ykakarap replied:
This diff was auto-generated as part of `make verify`.

BenTheElder replied:
cc @chaodaiG I would expect go version to be pinned in `make verify` (also `make verify` shouldn't mutate the source repo??) and verify in presubmit => no change to untouched files unless the PR touches .go-version. I think we have an issue somewhere then ...

chaodaiG replied:
afaict go version is pinned in `make verify`, but yes there seems like a few bugs, at least `make verify` should not mutate, please file an issue

michelle192837 commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -0,0 +1,1003 @@
+// knownApprovers must be a subset of potentialApprovers.
+func (o Owners) KeepCoveringApprovers(reverseMap map[string]sets.String, knownApprovals []Approval, potentialApprovers []string) sets.String {
+	if len(potentialApprovers) == 0 {
+		o.log.Debug("No potential approvers exist to filter for relevance. Does this repo have OWNERS files?")
```
nit: Not sure what "exist to filter for relevance" means.

ykakarap replied:
This is copied from the original implementation. This might be a good candidate for a follow-up PR that covers boarder log statements improvements.

michelle192837 commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -0,0 +1,1003 @@
+	for !ap.RequirementsMet() {
+		newApprover := findMostCoveringApprover(potentialApprovers, reverseMap, ap.UnapprovedFiles())
+		if newApprover == "" {
+			o.log.Warnf("Couldn't find/suggest approvers for each files. Unapproved: %q", ap.UnapprovedFiles().List())
```
nit: file

michelle192837 commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -0,0 +1,1003 @@
+	String() string
+}
+
+// ApprovedFolder contains the information of a an approved folder.
```
nit: (here and below) "a an" should be "an"

michelle192837 reviewed (commented):
Have some nitpicks, but I think in general it looks fine.

spiffxp commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -531,85 +604,150 @@ func (ap Approvers) RequirementsMet() bool {
+	if !reqsMet && ap.ManuallyApproved() {
+		return true
+	}
+	return reqsMet
```
I find the single line of code easier to understand, why the replacement?

spiffxp commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -725,3 +879,144 @@ func getGubernatorMetadata(toBeAssigned []string) string {
 }
+
+// wildcardPathMatch return true if the targetPath matches the given pattern
+// A '*' as a directory will recursively match all sub directories
+//    Example: file a/b1/b2/c.go matches the pattern a/*/c.go
```
This is surprising to me, not sure I like the syntax. It doesn't match https://pkg.go.dev/path/filepath#Match behavior.  I'm used to `**` signifying recursive matching (though go doesn't natively support this)

ykakarap replied:
The initial implementation did the `**` approach. Some folks suggested that using `*` might be easier instead of having to switch between `*` for any match and `**` for recursive match. 
cc @nikhita 

I am +1 to switching back if that is the majority's preference.

nikhita replied:
> The initial implementation did the ** approach. Some folks suggested that using * might be easier instead of having to switch between * for any match and ** for recursive match.

Honestly, I can't recall why we had decided on the `*` approach. If `**` works, let's just go with it. Would love to merge and iterate.

spiffxp replied:
I would prefer `**`

In terms of merge and iterate, I would prefer merging a PR that _just_ does per-file checking, and then we can iterate with a later PR that adds `/approve files foo` support.  Otherwise, I would rather see the `**` syntax fixed first.

spiffxp reviewed (changes requested):
I'm going to cut my review off here.  I want this to move forward, but I can already see a few things that need to change, and this is hard for me to digest from a review perspective.  I will try to review the actual test cases in another pass.

It would be really helpful if this was broken down into more commits, e.g.
- remove unused things from approve
- support regex-based owners files (common interface between approvers/approvers2 so same test cases/driver can be used in approve_test.go)
- add `/approve files foo` syntax

ykakarap commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -67,59 +59,67 @@ type Owners struct {
 }
 
-// GetApprovers returns a map from ownersFiles -> people that are approvers in them
+// GetApprovers returns a map from filenames -> people that can approve it
```
Yes this changes the logic to look at each file rather than each directory.

>What is the performance impact of this? e.g. any more github api calls than before?

I did some test on some really large PRs (~1200 files changes, ~500k lines modified) to see if there was any impact. I did not notice any significant slow down on impact when running these tests (I setup a separate prow instance to test this - I dont have those numbers with me now). 

I dont believe this makes any additional github calls. It uses the same information that the approval plugin receives. 

>How would this perform against some of the gnarlier size/XXL PR's we've seen? e.g. would the approvalnotifier comment be truncated?

In the new approval notifier message we only mention the count of the files and the list directories that have file changes. We do not list all the individual files. With this the message size does not grow a lot from the previous implementation and there is no truncation.

ykakarap commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -46,18 +46,10 @@ type Repo interface {
 	FindApproverOwnersForFile(file string) string
 	IsNoParentOwners(path string) bool
-	IsAutoApproveUnownedSubfolders(directory string) bool
-	Filenames() ownersconfig.Filenames
```
I think this is already addressed. It was probably just missed in a rebase. I will take a closer look again.

ykakarap commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -46,18 +46,10 @@ type Repo interface {
 	LeafApprovers(path string) sets.String
 	FindApproverOwnersForFile(file string) string
 	IsNoParentOwners(path string) bool
-	IsAutoApproveUnownedSubfolders(directory string) bool
```
I think this was resolved after doing a rebase.

ykakarap replied:
This was dropped in the rebase. Added the support back into the PR. Ref: https://github.com/kubernetes/test-infra/pull/21398/commits/4849287605277a9fab7aab2907c0816d5fdf4e6c

Note that additional logic was added to properly calculate the "target files" for approval calculations. Now, if any of the files in the PR are auto approved by the new setting then the approval notifier comment will consider them as approved in its message. Example: Consider a PR with 10 files out of which 4 are auto-approved with the new flag. 
The message will  read:
```
Out of 10 files: 4 are approved and 6 are unapproved.  
```

spiffxp replied:
See my latest review comments, I don't think this has been added back fully (and lacks test cases to confirm it works as expected)

MadhavJivrajani replied:
We’ll definitely need to add in tests to confirm this aspect of the functionality.

ykakarap commented on prow/plugins/approve/approve.go:
```diff
@@ -730,3 +663,241 @@ func getLast(cs []*comment) *comment {
+		if name == approveCommand {
+			ga.approversHandler.AddApprover(c.Author, c.HTMLURL, path)
+		} else {
+			ga.approversHandler.AddLGTMer(c.Author, c.HTMLURL, path)
```
Sure. That would be a good idea. Let's drop the arg-handling for `/lgtm` in a follow-up PR.

ykakarap commented on prow/plugins/approve/approve_test.go:
```diff
@@ -1367,6 +1367,1991 @@ Approvers can cancel approval by writing ` + "`/approve cancel`" + ` in a commen
+	tests := []struct {
+		name            string
+		branch          string
+		prAuthor        string
```
I believe this was done to test the cases where the PR was initially approved and properly cover the cases where the PR author is also in the OWNERS file.

ykakarap commented on prow/plugins/approve/approve_test.go:
```diff
@@ -1367,6 +1367,1991 @@ Approvers can cancel approval by writing ` + "`/approve cancel`" + ` in a commen
+		hasLabel        bool
+		humanApproved   bool
+		files           []string
+		approvers       map[string]layeredsets.String
+		leafApprovers   map[string]sets.String
+		approversOwners map[string]string
```
This was done so that we can get richer test cases where we can cover wider cases of file changes, folder structures and approvers.

ykakarap commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -404,45 +477,34 @@ func (ap Approvers) GetNoIssueApproversSet() sets.String {
-		filesApprovers[ownersFilename] = CaseInsensitiveIntersection(currentApprovers, potentialApprovers)
+	for fn, potentialApprovers := range ap.owners.GetApprovers() {
+		// potentialApprovers is the list of github handles who can approve file fn.
+		filesApprovers[fn] = approversForFile(fn, potentialApprovers, ap.approvers)
```
The `approversForFile` already takes care of case sensitivity. We don't directly take the intersection because we also need to check if the approval actually covers the file.

spiffxp replied:
OK I think you're right, this was just a very specific comment you deleted so I wanted to make sure you're still handling this corner case (mis-cased OWNERS file vs. github login)

ykakarap commented on prow/plugins/approve/approve.go:
```diff
@@ -730,3 +663,241 @@ func getLast(cs []*comment) *comment {
+		return
+	}
+
+	if reviewActsAsApprove && c.ReviewState == github.ReviewStateApproved {
+		ga.approversHandler.AddApprover(
+			c.Author,
+			c.HTMLURL,
+			"",
+		)
+	}
+	if reviewActsAsApprove && c.ReviewState == github.ReviewStateChangesRequested {
+		ga.approversHandler.RemoveApprover(c.Author)
+	}
+
+	for _, match := range commandRegex.FindAllStringSubmatch(c.Body, -1) {
```
Since the approvals in the new approve logic are appended (a user can give multiple approvals approving different sections of the PR) the order or processing comments and github reviews does not matter.

If a user submit a github review with "approve" state it just acts as a blanket approval. We should document this behavior in the plugin docs. (Here: https://github.com/kubernetes-sigs/prow/pull/16/files ?)

spiffxp replied:
>If a user submit a github review with "approve" state it just acts as a blanket approval. We should document this behavior in the plugin docs.

No.  The approve plugin doesn't work this way today.  Note "approval commands override GitHub review state" https://github.com/kubernetes/test-infra/blob/f24733b609a6d2b8ce9393a8a1eaaf88d7bbb592/prow/plugins/approve/approve.go#L243-L249

As Ben said, it would be unintuitive if turning on `granular_approval` changed this behavior.

If I submit an "approved" review with a comment of `/approve files foo`, it should only approve `foo`, and not be treated as a blanket approval

MadhavJivrajani replied:
@spiffxp -- working on the remaining comments, but wanted some clarification on this one.

The logic here is the same as the current approve plugin. Instead of processing the logic in `addApprovers` as here:
https://github.com/kubernetes/test-infra/blob/f24733b609a6d2b8ce9393a8a1eaaf88d7bbb592/prow/plugins/approve/approve.go#L586

the logic is refactored into a new function called `processComment`. If at all an approver gives an "approve review" with a `/approve files foo` comment, the `handleReviewEvent` event handler should let the `genericCommentEventHandler` handle it like you linked above. 

I can add a test for this case in the follow-up commits too.
Please lmk if I'm misunderstanding something here.

ykakarap commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -183,26 +185,31 @@ func (o Owners) GetSuggestedApprovers(reverseMap map[string]sets.String, potenti
-		} else {
-			owners.Insert(o.repo.FindApproverOwnersForFile(toApprove))
-		}
+	for _, fn := range o.filenames {
+		owners.Insert(o.repo.FindApproverOwnersForFile(fn))
```
I am afraid I dont follow what you are implying here. The function still returns the list of onwers files that are needed to cover the full PR (not a subset of the PR).

spiffxp replied:
I don't know either anymore, let's chalk it up to being lost in the review. You're right, the `removeSubdirs` call below turns this into the minimal covering set, so the comment is correct

ykakarap commented:
I addressed some of the comments. The broader items left that need to be address are:
* Break the commits further into logical block 
* Refactor some of the tests
* Support IsAutoApproveUnownedSubfolders in approvers2

alvaroaleman commented:
/uncc

ykakarap commented on prow/plugins/approve/approve.go:
```diff
@@ -730,3 +663,241 @@ func getLast(cs []*comment) *comment {
 	return cs[len(cs)-1]
 }
+
+type approveHelper interface {
```
Can you please elaborate more on how you are suggesting we do this?

spiffxp replied:
One option is to standardize on something like `Add(LGTMer|Approver)(login, reference string, noIssue bool, path string)` for both `Approvers` implementations, and ignore `path` in the original implementation

spiffxp commented on prow/plugins/approve/approve_test.go:
```diff
@@ -1372,6 +1372,1991 @@ Approvers can cancel approval by writing ` + "`/approve cancel`" + ` in a commen
 	}
 }
 
+func TestHandle_GanularApprover(t *testing.T) {
```
>It is difficult to tell that granular still handles everything simple does, and also to tell what granular handles that simple does not.
>
>It would be better if we could abstract to have a common testApprove(T, handler) type driver, or at least some commonality in fixtures

This still needs addressing

spiffxp commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -0,0 +1,1024 @@
+	owners := sets.NewString()
+	for _, fn := range o.filenames {
+		ownersFile := o.repo.FindApproverOwnersForFile(fn)
+		// If the ownersfile for fn is in the parent folder and has AllowFolderCreation enabled, we purge
+		// the file from our filenames list, because it doesn't need approval
```
I think the `IsAutoApproveUnownedSubfolders` commit is incomplete

The approvers2 `GetOwnersSet` implementation isn't doing what the comment says.  I think it should. In other words this should be a copy-paste of https://github.com/kubernetes/test-infra/blob/f24733b609a6d2b8ce9393a8a1eaaf88d7bbb592/prow/plugins/approve/approvers/owners.go#L183-L202

spiffxp commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -46,6 +46,7 @@ type Repo interface {
 	LeafApprovers(path string) sets.String
 	FindApproverOwnersForFile(file string) string
 	IsNoParentOwners(path string) bool
+	IsAutoApproveUnownedSubfolders(directory string) bool
```
Did you add (back) test cases that confirm this works as expected?

spiffxp commented on prow/plugins/approve/approvers2/owners.go:
```diff
@@ -46,18 +46,10 @@ type Repo interface {
-	filenamesUnfiltered []string
-	// filenames refers to the files in a given PR, not to OWNERS files. Files
-	// that are a directory below an Owners file with IsAutoApproveUnownedSubfolders
-	// are excluded here but kept in filenamesUnfiltered.
```
This remains unaddressed

MadhavJivrajani commented:
@spiffxp 👋🏼 

I've worked on your review comments and tried to keep the commits as small as independent as possible for ease of review. Here is a summary mapping what commits map to which review comments and what the purpose of commits that lie outside this mapping is:

- `fileNamesUnfiltered` should be in a separate commit: https://github.com/kubernetes/test-infra/pull/21398#discussion_r943018706
  - Commit: https://github.com/kubernetes/test-infra/pull/21398/commits/2a0209e7b714d440be392a0c448973cd8a634022
  - Note: This is re-added later as part of a different review comment, will describe below.
- `IsAutoApproveUnknownSubfolders` is not fully added back yet: https://github.com/kubernetes/test-infra/pull/21398#discussion_r1000020172
  - Commits:
    - Re-add the implementation https://github.com/kubernetes/test-infra/pull/21398/commits/a1e6146a8ca8e14cb3ad81cb05b1fda9f5a580ad
    -  Re-add original tests to test with granular approval: https://github.com/kubernetes/test-infra/pull/21398/commits/a57ff07670ff33bca4adcb8e0a868f2c52d4889b
    - Add new tests specific to granular approval: https://github.com/kubernetes/test-infra/pull/21398/commits/b6f8d8d94d1cac3b01d7ed159d24b4b006e02722
  - `fileNamesUnfiltered` is added back as part of this to keep the implementation as consistent as possible.
- Order of things in which the approve commands are processed (state not overridden correctly by comment): https://github.com/kubernetes/test-infra/pull/21398#discussion_r947202827
  - Commit: https://github.com/kubernetes/test-infra/pull/21398/commits/bed5fb6169709ff9879c9e3d3d0efb6cbe9633d0
- Redundant boilerplate due to bridge between approve and approve2: https://github.com/kubernetes/test-infra/pull/21398#discussion_r947222177
  - Commit: https://github.com/kubernetes/test-infra/pull/21398/commits/1ac52297714bb6bf2e33abd60b90de3f8898e788
- Using `**` to indicate recursive matching: https://github.com/kubernetes/test-infra/pull/21398#discussion_r947264343
  - Commit: https://github.com/kubernetes/test-infra/pull/21398/commits/6b70454f0ebbde1e30213d07bb16fa8282e318b7
---
#### Interlude 1

While working on refactoring tests, I discovered a few pit-falls in the implementation and testing (or at least that's what my understanding is). Post fixing these, approve2 is compliant with approve except the message format. Here's a summary of these changes:
- `GetApprovers()` correction: https://github.com/kubernetes/test-infra/pull/21398/commits/ee02b8978944c2e9d3ce0c327ea59dfadd87603a
- `GetLeafApprovers()` correction: https://github.com/kubernetes/test-infra/pull/21398/commits/9902a80ad876c0a929ea8d9af7508558eb642539
- `NoIssueApproval`s were not taken into account while calculating state of approval: https://github.com/kubernetes/test-infra/pull/21398/commits/2f415e2b4066ee4aaa47cd36d49c3b7c46b90c79
- Small modification to the approver suggestion algorithm specific to approve2: https://github.com/kubernetes/test-infra/pull/21398/commits/f3d6d4c0e6d2be3deba211c1600016d8868dc33a
---
- Test refactoring: https://github.com/kubernetes/test-infra/pull/21398#discussion_r943009419
  - Commits:
    - Add helper function to unify tests of simple and granular approver: https://github.com/kubernetes/test-infra/pull/21398/commits/575f160eb56c9f4ff970459d3f6b51e07c0ed99f (this commit did not turn out as cleanly as I hoped it would, sorry about that!)
    - Test granular approval with test cases of simple approval to ensure it does everything simple approve does: https://github.com/kubernetes/test-infra/pull/21398/commits/98d30b9021a5e4d928f3009f128a2f995e333322
    - Remove test cases that are already present in baseline test from the granular approve test: https://github.com/kubernetes/test-infra/pull/21398/commits/1258aed48b5fcc2f3d3bff578439044df44f0261

---
#### Interlude 2

According to my understanding, `approve2` interpreted `no-issue` as something that only bypasses the no-issue requirement but does not actually approve anything. This is different from simple approve where `no-issue` also results in approval. I guess it might make sense to have it such that it does not approve anything since that is a blanket approval but considering that `no-issue` hasn't been used in a while (https://github.com/kubernetes/test-infra/pull/6373) and no repos use it either, I'd like to keep the behaviour consistent across granular and simple approve until we sunset it (if at all). 

Commits relating to this:
- Fix tests:
  - https://github.com/kubernetes/test-infra/pull/21398/commits/e3b9ffb08ed1603a8a2128889082fa9fe3ff9b9c
  - https://github.com/kubernetes/test-infra/pull/21398/commits/d2dc63199da00b1e11c956d4cafca8630209d45d
---

- Misc commits relating to changes to message format/tests added at a later point that need to adapt to the changes made here:
   - https://github.com/kubernetes/test-infra/pull/21398/commits/0b6c7d5d2952c5f4b928b92c4e7b4234537f1845
   - https://github.com/kubernetes/test-infra/pull/21398/commits/3fdd5ce6b9584528ae141961aaf77c3bc9b1d9e0


Please lmk what you think and thank you again for your detailed review!

MadhavJivrajani commented on config/prow/config.yaml:
```diff
@@ -944,7 +944,7 @@ presets:
     - name: REGISTRY
       value: capzci.azurecr.io
-    - name: LOCAL_ONLY # This variable is used for backwards compatability, can be removed once CAPZ v1.7.0+ is used in all jobs
+    - name: LOCAL_ONLY # This variable is used for backwards compatibility, can be removed once CAPZ v1.7.0+ is used in all jobs
```
I'm not sure how the lint test didn't fail when this was merged

droslean commented:
/retest

droslean commented:
@ykakarap What is the status of this PR?

MadhavJivrajani commented:
@droslean 👋🏼 
We've addressed any existing review comments and the only thing required now is reviewer bandwidth which is unfortunately and understandably a scarce resource at the moment :/

droslean commented:
> @droslean 👋🏼 We've addressed any existing review comments and the only thing required now is reviewer bandwidth which is unfortunately and understandably a scarce resource at the moment :/

The PR is conflicted. Please make sure that this is resolved, otherwise, there is no point in reviewing it.

pohly commented:
This would be useful to grant dep approvers the actual rights to approve dependency changes across the k/k repo (i.e. go.mod/go.sum) without also granting them approval rights for other files.

pohly commented:
/remove-lifecycle stale

pohly commented:
@ykakarap: What are the chances that this work continues?

As Ben said, "some people want this quite badly!"

MadhavJivrajani commented:
/close
